onnxscript/ir/passes/_pass_infra.py

```diff
@@ -97,76 +145,102 @@ def ensures(self, model: ir.Model) -> None:
         del model  # Unused
 
 
-class PassManager:
+class InPlacePass(PassBase):
+    """A pass that modifies the input model in place and returns it."""
+
+    @property
+    def in_place(self) -> bool:
+        return True
+
+    @property
+    def changes_input(self) -> bool:
+        return True
+
+
+class OutOfPlacePass(PassBase):
+    """A pass that returns a new model but does not modify the input model."""
+
+    @property
+    def in_place(self) -> bool:
+        return False
+
+    @property
+    def changes_input(self) -> bool:
+        return False
+
+
+class Sequential(PassBase):
+    """Run a sequence of passes in order."""
+
+    def __init__(self, *passes: PassBase):
+        if not passes:
+            raise ValueError("Sequential must take at least one pass")
+        self.passes = passes
+        self._in_place = all(pass_.in_place for pass_ in passes)
+        self._changes_input = self.passes[0].changes_input or self.passes[0].in_place
```

Comment:
Changes_input looks exactly the same as in_place currently.

Reply:
Do you mean in a PassBase? They will look different when a user changes them or when passes are composed in Sequential (here)

Reply:
But I don't see a setter though. And why is changes_inputs is decided by the first pass?

Reply:
The properties are declared at init time. Users can subclass PassBase and define them.

The reason changes_inputs by the first pass is that if the first pass is either in-place, or if it is not designed to be in-place but somehow changes the input (destructive), this pass sequence will change inputs.

The relationship of the two properties is:

in_place x changes_inputs => in place
not in_place x changes_inputs => destructive
in_place x not changes_inputs => not possible, invalid
not in_place x not changes_inputs => functional